File: backend/src/modules/auth/user_schemas.py

"""
用户管理模块 - Pydantic Schema（v1.2）
v1.2 变更：
- P2-4：移除 nickname 的 min_length / max_length，改由 Service 层统一
        校验并返回 90001（请求参数错误），与错误码体系保持一致。
v1.1 变更：
- P1-4：UserCreate.username 去除 min_length / max_length 限制
- P2-5：UserOut.created_at / updated_at 改为必填
"""
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field
class RoleBrief(BaseModel):
    """角色简要信息（用于用户对象内嵌）。"""
    id: int
    name: str
    code: str
class UserCreate(BaseModel):
    """
    创建用户请求。
    注意：
        username / password / nickname 的长度校验全部交由 Service 层完成，
        以便统一返回业务错误码，而不是 Pydantic 的 422/90004。
    """
    username: str = Field(..., description="用户名（长度 3-20 由 service 校验）")
    password: str = Field(..., description="密码（长度 6-20 由 service 校验）")
    nickname: str = Field(..., description="昵称（长度 1-50 由 service 校验）")
    email: Optional[str] = Field(None, max_length=100, description="邮箱")
    role_ids: List[int] = Field(default_factory=list, description="角色 ID 列表")
    status: int = Field(1, ge=0, le=1, description="状态：1-启用 0-禁用")
class UserUpdate(BaseModel):
    """更新用户请求（所有字段可选）。"""
    nickname: Optional[str] = Field(None, description="昵称（长度 1-50 由 service 校验）")
    email: Optional[str] = Field(None, max_length=100, description="邮箱")
    role_ids: Optional[List[int]] = Field(None, description="角色 ID 列表（全量覆盖）")
    status: Optional[int] = Field(None, ge=0, le=1, description="状态：1-启用 0-禁用")
class UserStatusUpdate(BaseModel):
    """启用/禁用请求。"""
    status: int = Field(..., ge=0, le=1, description="状态：1-启用 0-禁用")
class UserPasswordReset(BaseModel):
    """管理员重置用户密码请求。"""
    new_password: str = Field(..., description="新密码（长度 6-20 由 service 校验）")
class UserOut(BaseModel):
    """用户响应对象。"""
    id: int
    username: str
    nickname: str
    email: Optional[str] = None
    avatar: Optional[str] = None
    status: int
    roles: List[RoleBrief] = []
    created_at: datetime
    updated_at: datetime

File: backend-go/internal/modules/auth/user_schemas.go
package auth
// UserCreate 创建用户请求。
//
// 必要字段增加 `binding:"required"`，缺失时在 HTTP 绑定层返回 422 / 90004。
// 长度/格式校验由 Service 层完成，统一返回业务错误码。
//
// Status 说明：
//   - 使用 *int8，省略时由 Service 层默认为 1（启用），
//     与 Python Pydantic Field(1, ge=0, le=1) 一致；
//   - 增加 binding:"omitempty,min=0,max=1"，
//     非法值（如 5 / -1）在 HTTP 绑定层返回 422 / 90004；
//   - ⚠️ P1-C 边界：JSON null 会被视为"未提供"，与 Python 行为不一致，
//     详见文档 § 4.12。
type UserCreate struct {
	Username string  `json:"username" binding:"required"`
	Password string  `json:"password" binding:"required"`
	Nickname string  `json:"nickname" binding:"required"`
	Email    *string `json:"email"`
	RoleIDs  []uint  `json:"role_ids"`
	Status   *int8   `json:"status" binding:"omitempty,min=0,max=1"`
}
// UserUpdate 更新用户请求（所有字段可选）。
type UserUpdate struct {
	Nickname *string `json:"nickname"`
	Email    *string `json:"email"`
	RoleIDs  *[]uint `json:"role_ids"`
	Status   *int8   `json:"status" binding:"omitempty,min=0,max=1"`
}
// UserStatusUpdate 启用/禁用请求。
//
// Status 使用 *int8 + binding:"required"，
// 保证字段存在性由 HTTP 绑定层校验，handler 无需二次检查。
type UserStatusUpdate struct {
	Status *int8 `json:"status" binding:"required,min=0,max=1"`
}
// UserPasswordReset 管理员重置用户密码请求。
type UserPasswordReset struct {
	NewPassword string `json:"new_password" binding:"required"`
}
// RoleBrief 角色简要信息（用户对象内嵌）。
type RoleBrief struct {
	ID   uint   `json:"id"`
	Name string `json:"name"`
	Code string `json:"code"`
}
// UserOut 用户响应对象。
type UserOut struct {
	ID        uint        `json:"id"`
	Username  string      `json:"username"`
	Nickname  string      `json:"nickname"`
	Email     *string     `json:"email"`
	Avatar    *string     `json:"avatar"`
	Status    int8        `json:"status"`
	Roles     []RoleBrief `json:"roles"`
	CreatedAt string      `json:"created_at"`
	UpdatedAt string      `json:"updated_at"`
}
// UserListQuery 用户列表查询参数。
type UserListQuery struct {
	Page     int    `form:"page" binding:"omitempty,min=1"`
	PageSize int    `form:"page_size" binding:"omitempty,min=1,max=100"`
	Keyword  string `form:"keyword"`
	Status   *int8  `form:"status" binding:"omitempty,min=0,max=1"`
	RoleID   *uint  `form:"role_id"`
}
